File: src/autora/memory.py
"""What the agent knows, between sessions.

Three things live here, and they are deliberately the same shape:

  * **facts and preferences** -- who you are, how you like things done
  * **procedures** -- how to do something in a particular project, which is
    what a "skill" is once you stop capitalising it
  * **episodes** -- what happened, compressed, so a later session can say
    "we tried that in March and it broke the migration"

Two design choices carry most of the weight.

**Everything has provenance.** A record points back at the session and event
that justified it. A memory system without that becomes a junk drawer: claims
accumulate, nothing can be checked, and a wrong belief is indistinguishable
from a right one. Being able to ask "why do you think that?" -- and to delete
the answer -- is what makes the rest trustworthy.

**Nothing is loaded speculatively.** The whole store is searched, a handful of
records are injected, and the rest stays one tool call away. A memory system
that pastes everything it knows into every prompt has just moved the context
problem somewhere more expensive.

Storage is SQLite with an FTS5 index, which means no service to run, no vector
database to keep in sync, and a store the user can open with any SQLite client.
Lexical search first is a real choice, not a shortcut: for one person's memory
the recall problem is small, and a BM25 hit you can explain beats a cosine
distance you cannot. Embeddings are a later optimisation if recall proves
insufficient, not a starting assumption.
"""

from __future__ import annotations

import json
import os
import re
import sqlite3
import time
import uuid
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Iterable

KINDS = ("fact", "preference", "procedure", "episode")
#: A record starts provisional and is confirmed when it proves itself a second
#: time. An auto-written procedure that nothing has re-validated is a guess, and
#: a guess the agent follows confidently is worse than no memory at all.
STATUSES = ("provisional", "confirmed", "retired")

GLOBAL = "global"


def project_scope(workdir: str | Path) -> str:
    """Scope key for a project. `pytest -x` is a fact about one repo, not the world."""
    return f"project:{Path(workdir).resolve()}"


@dataclass
class Record:
    id: str
    kind: str
    scope: str
    title: str
    body: str
    tags: list[str] = field(default_factory=list)
    status: str = "provisional"
    pinned: bool = False
    source_session: str | None = None
    source_seq: int | None = None
    created: float = 0.0
    updated: float = 0.0
    uses: int = 0
    last_used: float | None = None
    superseded_by: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {**self.__dict__, "tags": list(self.tags)}

    def as_context(self) -> str:
        """One record, as the model reads it."""
        head = f"[{self.id}] {self.title}"
        if self.status == "provisional":
            head += " (provisional)"
        return f"{head}\n{self.body}"


_SCHEMA = """
CREATE TABLE IF NOT EXISTS records (
  id TEXT PRIMARY KEY,
  kind TEXT NOT NULL,
  scope TEXT NOT NULL,
  title TEXT NOT NULL,
  body TEXT NOT NULL,
  tags TEXT NOT NULL DEFAULT '',
  status TEXT NOT NULL DEFAULT 'provisional',
  pinned INTEGER NOT NULL DEFAULT 0,
  source_session TEXT,
  source_seq INTEGER,
  created REAL NOT NULL,
  updated REAL NOT NULL,
  uses INTEGER NOT NULL DEFAULT 0,
  last_used REAL,
  superseded_by TEXT
);
CREATE INDEX IF NOT EXISTS records_scope ON records(scope, status);
CREATE TABLE IF NOT EXISTS links (
  src TEXT NOT NULL, dst TEXT NOT NULL, rel TEXT NOT NULL DEFAULT 'related',
  PRIMARY KEY (src, dst, rel)
);
CREATE VIRTUAL TABLE IF NOT EXISTS records_fts
  USING fts5(title, body, tags, content='records', content_rowid='rowid');
-- Keep the index in step with the table. Triggers rather than application code
-- so a write through any other client cannot desynchronise the two.
CREATE TRIGGER IF NOT EXISTS records_ai AFTER INSERT ON records BEGIN
  INSERT INTO records_fts(rowid, title, body, tags)
  VALUES (new.rowid, new.title, new.body, new.tags);
END;
CREATE TRIGGER IF NOT EXISTS records_ad AFTER DELETE ON records BEGIN
  INSERT INTO records_fts(records_fts, rowid, title, body, tags)
  VALUES ('delete', old.rowid, old.title, old.body, old.tags);
END;
CREATE TRIGGER IF NOT EXISTS records_au AFTER UPDATE ON records BEGIN
  INSERT INTO records_fts(records_fts, rowid, title, body, tags)
  VALUES ('delete', old.rowid, old.title, old.body, old.tags);
  INSERT INTO records_fts(rowid, title, body, tags)
  VALUES (new.rowid, new.title, new.body, new.tags);
END;
"""

#: Words too common to narrow anything, plus the scaffolding of a request. They
#: are stripped from queries only -- the records keep their own text.
_STOP = frozenset("""
a an the and or but if then than that this these those is are was were be been
being do does did doing have has had having i me my we our you your it its of
to in on at for with from by as so no not can could should would will just
please help need want make made get got use used using how what why when where
which who whom about into over under again more most some any all both each
""".split())


def default_db(root: Path | None = None) -> Path:
    base = root or Path(os.environ.get("AUTORA_HOME", Path.home() / ".autora"))
    return Path(base) / "memory.db"


class MemoryStore:
    """The knowledge store. One file, openable by anything that speaks SQLite."""

    def __init__(self, path: Path | None = None):
        self.path = Path(path) if path else default_db()
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self.db = sqlite3.connect(str(self.path), check_same_thread=False)
        self.db.row_factory = sqlite3.Row
        self.db.executescript(_SCHEMA)
        self.db.commit()
        self._backfill_links()

    # -- writing ---------------------------------------------------------

    def write(
        self,
        kind: str,
        title: str,
        body: str,
        scope: str = GLOBAL,
        tags: Iterable[str] = (),
        status: str = "provisional",
        pinned: bool = False,
        source_session: str | None = None,
        source_seq: int | None = None,
        supersedes: str | None = None,
    ) -> Record:
        """Add a record, or confirm the one it repeats.

        Writing the same thing twice is the confirmation signal: a claim that
        turns up again under the same scope has earned more than provisional
        status. It also stops the store filling with near-duplicates, which is
        the usual way these things rot.
        """
        if kind not in KINDS:
            raise ValueError(f"kind must be one of {KINDS}, got {kind!r}")
        now = time.time()
        tags = [t.strip().lower() for t in tags if t and t.strip()]

        existing = self._same_as(title, scope)
        if existing is not None:
            merged = sorted(set(existing.tags) | set(tags))
            self.db.execute(
                "UPDATE records SET body=?, tags=?, status=?, updated=?, "
                "pinned=?, source_session=COALESCE(?, source_session), "
                "source_seq=COALESCE(?, source_seq) WHERE id=?",
                (body, " ".join(merged),
                 "confirmed" if existing.status != "retired" else existing.status,
                 now, int(pinned or existing.pinned), source_session, source_seq,
                 existing.id),
            )
            self.db.commit()
            return self.get(existing.id)  # type: ignore[return-value]

        record = Record(
            id=uuid.uuid4().hex[:10], kind=kind, scope=scope, title=title.strip(),
            body=body.strip(), tags=sorted(set(tags)), status=status, pinned=pinned,
            source_session=source_session, source_seq=source_seq,
            created=now, updated=now,
        )
        self.db.execute(
            "INSERT INTO records (id, kind, scope, title, body, tags, status, pinned,"
            " source_session, source_seq, created, updated) "
            "VALUES (?,?,?,?,?,?,?,?,?,?,?,?)",
            (record.id, record.kind, record.scope, record.title, record.body,
             " ".join(record.tags), record.status, int(record.pinned),
             record.source_session, record.source_seq, record.created, record.updated),
        )
        if supersedes:
            # A contradicted belief is retired, not deleted: the record of having
            # believed it is part of why the new one is trusted.
            self.db.execute(
                "UPDATE records SET status='retired', superseded_by=?, updated=? WHERE id=?",
                (record.id, now, supersedes))
            self.link(supersedes, record.id, "superseded-by")
        self.db.commit()
        # A record only becomes part of a web once something connects it.
        self.relate(record)
        return record

    def _backfill_links(self) -> None:
        """Infer links once, for a store that predates inferring them.

        Only when there are records and no links at all: that pairing means
        nobody has linked anything, which before this existed was every store.
        Guarded so it cannot run repeatedly, and skipped for large stores where
        a startup pass would be felt.
        """
        try:
            records = self.db.execute(
                "SELECT COUNT(*) FROM records WHERE status != 'retired'").fetchone()[0]
            links = self.db.execute("SELECT COUNT(*) FROM links").fetchone()[0]
        except sqlite3.Error:
            return
        if records == 0 or links > 0 or records > 300:
            return
        self.relate_all()

    def relate(self, record: Record, limit: int = 3) -> int:
        """Connect a record to the ones it is plainly about.

        Links were only ever created by an explicit request from the model or
        by a supersede, so two facts about the same subject sat side by side
        with nothing between them and the web was a scatter of unconnected
        dots. The relationship was real; nothing was writing it down.

        Relatedness is inferred from the same index that answers recall -- a
        shared significant word, within one scope -- so it is wrong in the same
        ways recall is wrong rather than in some new way of its own. A shared
        word is required rather than trusting the ranking alone: BM25 will
        always return something, and a graph that links everything to
        everything says as little as one that links nothing.
        """
        own = set(_terms(f"{record.title} {' '.join(record.tags)}"))
        if not own:
            return 0
        made = 0
        for hit in self.search(record.title, scopes=[record.scope], limit=limit + 5):
            if made >= limit:
                break
            if hit.id == record.id:
                continue
            if own & set(_terms(f"{hit.title} {' '.join(hit.tags)}")):
                self.link(record.id, hit.id, "related")
                made += 1
        return made

    def relate_all(self, cap: int = 300) -> int:
        """Infer links across the whole store.

        For records written before anything inferred links. Idempotent -- the
        links table ignores duplicates -- so running it twice costs time and
        changes nothing.
        """
        made = 0
        for record in self.all()[:cap]:
            made += self.relate(record)
        return made

    def link(self, src: str, dst: str, rel: str = "related") -> None:
        if src == dst:
            return
        self.db.execute("INSERT OR IGNORE INTO links (src, dst, rel) VALUES (?,?,?)",
                        (src, dst, rel))
        self.db.commit()

    def forget(self, record_id: str, hard: bool = False) -> bool:
        """Retire a record, or erase it outright.

        Retiring is the default because a wrong memory usually wants to stay
        visible as a wrong memory. Hard delete exists because some things should
        genuinely leave the disk when asked.
        """
        if hard:
            cur = self.db.execute("DELETE FROM records WHERE id=?", (record_id,))
            self.db.execute("DELETE FROM links WHERE src=? OR dst=?", (record_id, record_id))
        else:
            cur = self.db.execute(
                "UPDATE records SET status='retired', updated=? WHERE id=?",
                (time.time(), record_id))
        self.db.commit()
        return cur.rowcount > 0

    def touch(self, ids: Iterable[str]) -> None:
        """Record that these were actually used, which is what ranks them later."""
        now = time.time()
        self.db.executemany("UPDATE records SET uses = uses + 1, last_used=? WHERE id=?",
                            [(now, i) for i in ids])
        self.db.commit()

    # -- reading ---------------------------------------------------------

    def get(self, record_id: str) -> Record | None:
        row = self.db.execute("SELECT * FROM records WHERE id=?", (record_id,)).fetchone()
        return _row(row) if row else None

    def all(self, include_retired: bool = False) -> list[Record]:
        sql = "SELECT * FROM records"
        if not include_retired:
            sql += " WHERE status != 'retired'"
        return [_row(r) for r in self.db.execute(sql + " ORDER BY updated DESC")]

    def links(self) -> list[dict[str, str]]:
        return [dict(r) for r in self.db.execute("SELECT src, dst, rel FROM links")]

    def search(
        self,
        query: str,
        scopes: Iterable[str] = (),
        limit: int = 12,
        include_retired: bool = False,
    ) -> list[Record]:
        """Rank by relevance, then by how much the record has earned trust.

        BM25 gives the lexical match. The rest is a small, explainable nudge:
        pinned first, confirmed over provisional, and records that have actually
        been used over ones that never have. No magic numbers you cannot reason
        about at 3am.
        """
        terms = _terms(query)
        scopes = list(scopes) or [GLOBAL]
        if not terms:
            rows = self.db.execute(
                f"SELECT * FROM records WHERE scope IN ({_qs(scopes)})"
                + ("" if include_retired else " AND status != 'retired'")
                + " ORDER BY pinned DESC, uses DESC, updated DESC LIMIT ?",
                (*scopes, limit)).fetchall()
            return [_row(r) for r in rows]

        match = " OR ".join(f'"{t}"' for t in terms)
        sql = (
            "SELECT r.*, bm25(records_fts) AS rank FROM records_fts "
            "JOIN records r ON r.rowid = records_fts.rowid "
            f"WHERE records_fts MATCH ? AND r.scope IN ({_qs(scopes)})"
            + ("" if include_retired else " AND r.status != 'retired'")
            + " ORDER BY rank LIMIT ?"
        )
        rows = self.db.execute(sql, (match, *scopes, limit * 3)).fetchall()

        def score(row: sqlite3.Row) -> float:
            s = -float(row["rank"])                      # bm25 is negative-better
            if row["pinned"]:
                s += 6.0
            if row["status"] == "confirmed":
                s += 1.5
            s += min(row["uses"], 5) * 0.3
            return -s

        return [_row(r) for r in sorted(rows, key=score)[:limit]]

    def _same_as(self, title: str, scope: str) -> Record | None:
        row = self.db.execute(
            "SELECT * FROM records WHERE scope=? AND lower(title)=lower(?)",
            (scope, title.strip())).fetchone()
        return _row(row) if row else None

    def close(self) -> None:
        self.db.close()


def _qs(items: list[str]) -> str:
    return ",".join("?" * len(items))


def _row(row: sqlite3.Row) -> Record:
    data = dict(row)
    data.pop("rank", None)
    data["tags"] = [t for t in (data.get("tags") or "").split(" ") if t]
    data["pinned"] = bool(data["pinned"])
    return Record(**data)


def _terms(query: str) -> list[str]:
    """Query words worth searching on.

    Stopwords are dropped and FTS syntax is stripped: a user sentence is not a
    query language, and letting `NOT` or a stray quote through turns a search
    into a syntax error.
    """
    words = re.findall(r"[A-Za-z0-9_./-]{2,}", query.lower())
    return [w for w in dict.fromkeys(words) if w not in _STOP][:12]

File: server/memory.ts
/**
 * The memory graph: what the agent knows between sessions, and how it keeps
 * that knowledge from rotting.
 *
 * Three things this has to get right that a list of notes does not:
 *
 *  1. Recall by meaning, not by luck. Memories are ranked against the request
 *     (BM25 over title, tags and body), with a lift for ones that have been
 *     useful before, so a memory written yesterday is found as readily as one
 *     the app shipped with.
 *  2. No pile of near-copies. Writing something close to what is already
 *     there updates it instead of adding a second, slightly different one.
 *  3. Learned is not the same as known. What the agent works out for itself
 *     is `provisional` until it is confirmed -- by the person, or by working
 *     again -- and a provisional change to a confirmed memory is a new record
 *     that only replaces the old one once it is confirmed.
 *
 * The class holds the arrays and nothing else; the server owns saving them.
 */

export type MemoryKind = "fact" | "preference" | "procedure" | "skill";
export const MEMORY_KINDS: MemoryKind[] = ["preference", "procedure", "fact", "skill"];

export interface MemoryRecord {
  id: string;
  kind: MemoryKind;
  scope: string;
  title: string;
  body: string;
  tags: string[];
  status: "provisional" | "confirmed";
  pinned: boolean;
  source_session: string | null;
  source_seq: number | null;
  created: number;
  updated: number;
  /** Times it was put in front of the model. */
  uses: number;
  last_used: number | null;
  superseded_by: string | null;
  /** Times a turn that used it went well. Drives confirming and promoting. */
  worked?: number;
  /** A provisional rewrite of this record, which replaces it on confirm. */
  replaces?: string | null;
}

export interface MemoryLink {
  src: string;
  dst: string;
  rel: string;
}

export interface Recalled {
  record: MemoryRecord;
  score: number;
  /** Why it was picked, in words for the chip's tooltip. */
  reason: string;
}

/** A provisional memory that has worked this many times is confirmed. */
export const CONFIRM_AFTER = 2;
/** A confirmed procedure that has worked this many times is marked proven,
    which ranks it higher whenever it matches. It is not pinned: see reinforce. */
export const PROMOTE_AFTER = 3;
/** Unconfirmed, unused memories older than this are dropped by consolidate. */
const STALE_PROVISIONAL_S = 30 * 24 * 3600;

// ------------------------------------------------------------------ text --

const STOP = new Set(
  ("a an and are as at be but by can do does for from has have how i if in into is it its " +
    "me my of on or our so that the their them then there these this to use was we what " +
    "when where which who why will with you your please just also should would could about " +
    "all any some get got make made want need").split(" "),
);

/** Lowercase words, stop words out, crude plural and tense endings off. */
export function tokens(text: string): string[] {
  return (text.toLowerCase().match(/[a-z0-9][a-z0-9_.-]*[a-z0-9]|[a-z0-9]/g) ?? [])
    .filter((w) => !STOP.has(w))
    .map(stem);
}

function stem(word: string): string {
  if (word.length > 5 && word.endsWith("ing")) return word.slice(0, -3);
  if (word.length > 4 && word.endsWith("ed")) return word.slice(0, -2);
  if (word.length > 4 && word.endsWith("es")) return word.slice(0, -2);
  if (word.length > 3 && word.endsWith("s") && !word.endsWith("ss")) return word.slice(0, -1);
  return word;
}

/** Overlap of two texts' vocabularies, 0..1. */
export function similarity(a: string, b: string): number {
  const x = new Set(tokens(a));
  const y = new Set(tokens(b));
  if (x.size === 0 || y.size === 0) return 0;
  let shared = 0;
  for (const w of x) if (y.has(w)) shared += 1;
  return shared / (x.size + y.size - shared);
}

const now = () => Math.floor(Date.now() / 1000);
const GENERIC_TAGS = new Set(["skill", "agent-authored", "learned", "unconfirmed", "proven"]);

/**
 * Whether a memory that shares `hits` of the request's `asked` words is about
 * the request rather than merely mentioning one of them.
 *
 * One shared word is enough only when it is what the memory is about (in its
 * title or tags) or the request is a word or two long. Otherwise a memory has
 * to share two words with it -- a single "file" or "page" in a long body is
 * how unrelated skills used to ride along on every turn.
 */
function relevant(hits: number, headHits: number, asked: number): boolean {
  if (hits >= 2) return true;
  return headHits >= 1 || asked <= 2;
}

// ----------------------------------------------------------------- graph --

export class MemoryGraph {
  constructor(
    public readonly records: MemoryRecord[],
    public readonly links: MemoryLink[],
    private readonly changed: () => void = () => undefined,
  ) {
    /* Proven procedures used to be pinned, and so put in front of the model
       on every turn whatever it was asked -- a pile of unrelated skills on a
       question about the weather. Undo that for the ones promoted before. */
    let unpinned = 0;
    for (const r of records) {
      if (r.pinned && r.kind === "procedure" && r.tags.includes("proven")) {
        r.pinned = false;
        unpinned += 1;
      }
    }
    if (unpinned) this.changed();
  }

  /** Everything still current: superseded records are history. */
  active(): MemoryRecord[] {
    return this.records.filter((r) => !r.superseded_by);
  }

  get(id: string): MemoryRecord | undefined {
    return this.records.find((r) => r.id === id);
  }

  /**
   * The memories that bear on `query`, best first.
   *
   * Pinned memories -- only ever pinned by the person -- are always included
   * unless `withPinned` is false. The rest must be about the query, not just
   * share a word with it: see `relevant`. `limit` caps how many come back.
   */
  recall(query: string, limit = 6, withPinned = true): Recalled[] {
    const pool = this.active();
    const want = [...new Set(tokens(query))];
    const docs = pool.map((r) => {
      // Bookkeeping tags ("skill", "learned") say how a memory came to be,
      // not what it is about; matched, they would recall every skill at once.
      const topical = r.tags.filter((t) => !GENERIC_TAGS.has(t)).flatMap((t) => tokens(t));
      const head = new Set([...tokens(r.title), ...topical]);
      // Title and tags count double: they are what the memory is about.
      const words = [...tokens(r.title), ...tokens(r.title), ...topical.flatMap((t) => [t, t]), ...tokens(r.body)];
      // Counted once here rather than by scanning every word of every memory
      // for every word of the query, which is what recall did on each turn.
      const tf = new Map<string, number>();
      for (const w of words) tf.set(w, (tf.get(w) ?? 0) + 1);
      return { record: r, head, words, tf };
    });
    const avg = docs.reduce((n, d) => n + d.words.length, 0) / Math.max(docs.length, 1) || 1;
    const df = new Map<string, number>();
    for (const d of docs) for (const w of d.tf.keys()) df.set(w, (df.get(w) ?? 0) + 1);

    const scored: Recalled[] = [];
    for (const d of docs) {
      let score = 0;
      const hit: string[] = [];
      let headHits = 0;
      for (const w of want) {
        const tf = d.tf.get(w) ?? 0;
        if (!tf) continue;
        hit.push(w);
        if (d.head.has(w)) headHits += 1;
        const idf = Math.log(1 + (docs.length - (df.get(w) ?? 0) + 0.5) / ((df.get(w) ?? 0) + 0.5));
        score += idf * ((tf * 2.2) / (tf + 1.2 * (0.25 + 0.75 * (d.words.length / avg))));
      }
      if (score > 0 && relevant(hit.length, headHits, want.length)) {
        const r = d.record;
        score *= 1 + 0.1 * Math.log1p(r.uses) + 0.15 * Math.log1p(r.worked ?? 0);
        if (r.tags.includes("proven")) score *= 1.2;
        if (r.status === "provisional") score *= 0.8;
        scored.push({ record: r, score, reason: `matched ${hit.slice(0, 4).join(", ")}` });
      }
    }
    scored.sort((a, b) => b.score - a.score);

    const out: Recalled[] = pool
      .filter((r) => withPinned && r.pinned)
      .map((record) => ({ record, score: Infinity, reason: "pinned: always recalled" }));
    const pinned = out.length;
    // A long tail of weak matches is noise: keep what scores within reach
    // of the best.
    const floor = (scored[0]?.score ?? 0) * 0.4;
    for (const s of scored) {
      if (out.length >= pinned + limit || s.score < floor) break;
      if (!out.some((o) => o.record.id === s.record.id)) out.push(s);
    }
    return out;
  }

  /** Counted as used: it was put in front of the model. */
  touch(ids: string[]) {
    const t = now();
    for (const id of ids) {
      const r = this.get(id);
      if (!r) continue;
      r.uses += 1;
      r.last_used = t;
    }
    if (ids.length) this.changed();
  }

  /** The existing memory this would be a near-copy of, if any. */
  findDuplicate(title: string, body: string, kind: MemoryKind): MemoryRecord | null {
    const norm = (s: string) => tokens(s).join(" ");
    let best: MemoryRecord | null = null;
    let bestSim = 0;
    for (const r of this.active()) {
      if (r.kind !== kind) continue;
      if (norm(r.title) && norm(r.title) === norm(title)) return r;
      const sim = similarity(`${r.title} ${r.body}`, `${title} ${body}`);
      if (sim > bestSim) { bestSim = sim; best = r; }
    }
    return bestSim >= 0.55 ? best : null;
  }

  /**
   * Write something down.
   *
   * `confirmed` writes (the agent's own memory_write, or the person) update a
   * near-duplicate in place. `provisional` ones (learned after a turn) never
   * overwrite a confirmed memory: they become a candidate that replaces it
   * only when confirmed, and one that says the same thing again only counts
   * as the old one having worked.
   */
  write(input: {
    title: string;
    body: string;
    kind?: string;
    tags?: string[];
    status?: "confirmed" | "provisional";
    source_session?: string | null;
    source_seq?: number | null;
  }): { record: MemoryRecord; action: "added" | "merged" | "reinforced" | "proposed" } {
    const kind: MemoryKind = MEMORY_KINDS.includes(input.kind as MemoryKind) ? (input.kind as MemoryKind) : "fact";
    const status = input.status ?? "confirmed";
    const title = input.title.trim().slice(0, 200);
    const body = input.body.trim();
    const tags = [...new Set((input.tags ?? []).map((t) => String(t).trim().toLowerCase()).filter(Boolean))];
    const t = now();

    const dup = this.findDuplicate(title, body, kind);
    if (dup && (status === "confirmed" || dup.status === "provisional")) {
      dup.title = title || dup.title;
      dup.body = body || dup.body;
      dup.tags = [...new Set([...dup.tags, ...tags])];
      dup.updated = t;
      if (status === "confirmed") this.confirm(dup.id, false);
      this.changed();
      return { record: dup, action: "merged" };
    }
    if (dup && similarity(dup.body, body) >= 0.8) {
      this.reinforce(dup.id);
      return { record: dup, action: "reinforced" };
    }

    const record: MemoryRecord = {
      id: `mem-${t.toString(36)}${Math.random().toString(36).slice(2, 6)}`,
      kind,
      scope: "workspace",
      title,
      body,
      tags: status === "provisional" ? [...new Set([...tags, "learned"])] : tags,
      status,
      pinned: false,
      source_session: input.source_session ?? null,
      source_seq: input.source_seq ?? null,
      created: t,
      updated: t,
      uses: 0,
      last_used: null,
      superseded_by: null,
      worked: 0,
      replaces: dup ? dup.id : null,
    };
    this.records.push(record);
    if (dup) this.links.push({ src: record.id, dst: dup.id, rel: "revises" });
    this.linkRelated(record);
    this.changed();
    return { record, action: dup ? "proposed" : "added" };
  }

  /** Tie a new memory to the few it most resembles, so the graph grows. */
  linkRelated(record: MemoryRecord, max = 3) {
    const mine = new Set(record.tags.filter((t) => !GENERIC_TAGS.has(t)));
    const candidates = this.active()
      .filter((r) => r.id !== record.id && r.id !== record.replaces)
      .map((r) => {
        const sharedTags = r.tags.filter((t) => mine.has(t)).length;
        const sim = similarity(`${r.title} ${r.body}`, `${record.title} ${record.body}`);
        return { r, weight: sim + 0.15 * sharedTags };
      })
      .filter((c) => c.weight >= 0.18)
      .sort((a, b) => b.weight - a.weight)
      .slice(0, max);
    for (const { r } of candidates) {
      if (this.links.some((l) => (l.src === record.id && l.dst === r.id) || (l.src === r.id && l.dst === record.id))) continue;
      this.links.push({ src: record.id, dst: r.id, rel: "related" });
    }
  }

  update(id: string, patch: { title?: string; body?: string; kind?: string; tags?: string[]; pinned?: boolean }): MemoryRecord | null {
    const r = this.get(id);
    if (!r) return null;
    if (typeof patch.title === "string" && patch.title.trim()) r.title = patch.title.trim();
    if (typeof patch.body === "string" && patch.body.trim()) r.body = patch.body.trim();
    if (MEMORY_KINDS.includes(patch.kind as MemoryKind)) r.kind = patch.kind as MemoryKind;
    if (Array.isArray(patch.tags)) r.tags = patch.tags.map((t) => String(t).trim()).filter(Boolean);
    if (typeof patch.pinned === "boolean") r.pinned = patch.pinned;
    r.updated = now();
    this.changed();
    return r;
  }

  /** Retire a memory: replaced by another if one is named, else removed. */
  forget(id: string, replacedBy?: string | null): boolean {
    const r = this.get(id);
    if (!r) return false;
    // A replacement that does not exist is a mistake, not a request to delete.
    if (replacedBy && (!this.get(replacedBy) || replacedBy === id)) return false;
    if (replacedBy) {
      r.superseded_by = replacedBy;
      r.pinned = false;
      r.updated = now();
    } else {
      this.records.splice(this.records.indexOf(r), 1);
      for (let i = this.links.length - 1; i >= 0; i -= 1) {
        if (this.links[i].src === id || this.links[i].dst === id) this.links.splice(i, 1);
      }
      for (const other of this.records) if (other.replaces === id) other.replaces = null;
    }
    this.changed();
    return true;
  }

  /** Make a provisional memory known; a rewrite then retires what it rewrote. */
  confirm(id: string, save = true): MemoryRecord | null {
    const r = this.get(id);
    if (!r) return null;
    r.status = "confirmed";
    r.tags = r.tags.filter((t) => t !== "learned" && t !== "unconfirmed");
    if (r.replaces) {
      const old = this.get(r.replaces);
      if (old && !old.superseded_by) {
        old.superseded_by = r.id;
        if (old.pinned) r.pinned = true;
        old.pinned = false;
      }
      r.replaces = null;
    }
    r.updated = now();
    if (save) this.changed();
    return r;
  }

  /**
   * A turn that used this memory went well.
   *
   * Enough of those confirm a provisional memory without anyone having to,
   * and mark a confirmed procedure proven, which ranks it higher when it is
   * relevant. It is never pinned for it: always-recalled is for what the
   * person pins, since a skill that worked for one job is noise on every
   * other.
   * Returns what changed, for the thread.
   */
  reinforce(id: string): "confirmed" | "promoted" | null {
    const r = this.get(id);
    if (!r || r.superseded_by) return null;
    r.worked = (r.worked ?? 0) + 1;
    let change: "confirmed" | "promoted" | null = null;
    if (r.status === "provisional" && r.worked >= CONFIRM_AFTER) {
      this.confirm(r.id, false);
      change = "confirmed";
    } else if (
      r.status === "confirmed" && r.kind === "procedure" && !r.tags.includes("proven") &&
      r.worked >= PROMOTE_AFTER
    ) {
      r.tags = [...new Set([...r.tags, "proven"])];
      change = "promoted";
    }
    this.changed();
    return change;
  }

  /**
   * Housekeeping, run daily: merge near-copies that slipped in, drop learned
   * guesses nobody confirmed or used within a month, and drop links to
   * records that are gone.
   */
  consolidate(at = now()): { merged: number; dropped: number; unlinked: number } {
    let merged = 0;
    let dropped = 0;
    const current = this.active();
    for (let i = 0; i < current.length; i += 1) {
      const a = current[i];
      if (a.superseded_by) continue;
      for (let j = i + 1; j < current.length; j += 1) {
        const b = current[j];
        if (b.superseded_by || a.kind !== b.kind || a.status !== b.status) continue;
        if (similarity(`${a.title} ${a.body}`, `${b.title} ${b.body}`) < 0.8) continue;
        const [keep, old] = a.updated >= b.updated ? [a, b] : [b, a];
        keep.uses += old.uses;
        keep.worked = (keep.worked ?? 0) + (old.worked ?? 0);
        keep.pinned = keep.pinned || old.pinned;
        keep.tags = [...new Set([...keep.tags, ...old.tags])];
        old.superseded_by = keep.id;
        old.pinned = false;
        merged += 1;
      }
    }
    for (const r of [...this.records]) {
      if (
        r.status === "provisional" && !r.pinned && (r.worked ?? 0) === 0 && r.uses <= 1 &&
        at - r.created > STALE_PROVISIONAL_S
      ) {
        this.records.splice(this.records.indexOf(r), 1);
        dropped += 1;
      }
    }
    const ids = new Set(this.records.map((r) => r.id));
    const before = this.links.length;
    for (let i = this.links.length - 1; i >= 0; i -= 1) {
      if (!ids.has(this.links[i].src) || !ids.has(this.links[i].dst)) this.links.splice(i, 1);
    }
    const unlinked = before - this.links.length;
    if (merged || dropped || unlinked) this.changed();
    return { merged, dropped, unlinked };
  }
}
